### DISCUSSION TITLE
Form submitted from within a modal - how to close the modal AND reload the page?

### DISCUSSION THREAD

**Discussion by vladstudio** [unpoly/unpoly#347] [Q&A] [answered]
Sorry guys, I'm lost (again).
So I'm developing a website, with a common header, containing link to Log in (opens a modal with login form). After user has submitted the form and server validated it and created a user cookie, how do I (1) close the modal, (2) reload the page underneath the modal? So far I was able to respond from server with

`header('X-Up-Dismiss-Layer: null');`

which closes the modal just fine, but I cannot find how do I reload the root (the entire page) at the same time. Thanks!

**Comment by triskweline** [comment, marked as answer]
There are [many ways to do this](https://unpoly.com/closing-overlays).

Maybe the shortest solution would be to make the login form within the overlay target the root layer:

```ruby
<!-- form within modal -->
<form action="/login" up-layer="root">
  <input type="email" ...>
  <input type="password" ...>
  <button type="submit">Login</button>
</form>
```

Login failures (server sends non-2xx HTTP status) would still be shown in the modal layer.

Another solution would be to tell the link that opens the modal overlay:

- when is the login sub-interaction done ("when we reach the /welcome path")
- what to do when the sub-interaction is done ("reload the layer")

```ruby
<a href="/login" up-layer="new" up-accept-location="/welcome" up-on-accepted="up.reload()">
  Login
</a>
```

> `header('X-Up-Dismiss-Layer: null');`

Unpoly distinguishes two different [close intents](https://unpoly.com/closing-overlays#distinguishing-close-intents):

- Accepting an overlay (user picks a value, confirms with "OK", etc.), optionally with a value
- Dismissing an overlay (user clicks "Cancel", "X", presses ESC, clicks on the background)

In this case *accepting* would be the canonical choice, so I chose that in the code examples above.

**Reply by vladstudio** [reply]
That is incredibly userful - thanks!

**Reply by vladstudio** [reply]
For reference: in my case, it was enough to send `header('X-Up-Accept-Layer: null');` from server, and modify the link to `<a ... up-layer="new modal" up-on-accepted="up.reload()">`
